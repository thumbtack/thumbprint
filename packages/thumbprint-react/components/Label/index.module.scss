@import '~@thumbtack/thumbprint-tokens/dist/scss/_index';

.text {
    display: flex;
    align-items: center;
    font-weight: 700;
    font-size: $tp-font__body__1__size;
    line-height: $tp-font__body__1__line-height;
    margin-bottom: $tp-space__2;

    &UiState {
        &Disabled {
            color: $tp-color__gray;
        }

        &Error {
            color: $tp-color__red;
        }

        &Default {
            color: $tp-color__black;
        }
    }
}
