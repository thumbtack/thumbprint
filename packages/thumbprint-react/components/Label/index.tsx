import React from 'react';
import classNames from 'classnames';
import styles from './index.module.scss';

type UiState = 'disabled' | 'error' | 'default';

const getUIState = ({
    isDisabled,
    hasError,
}: Pick<PropTypes, 'isDisabled' | 'hasError'>): UiState => {
    if (isDisabled) {
        return 'disabled';
    }

    if (hasError) {
        return 'error';
    }

    return 'default';
};

interface PropTypes {
    /**
     * The label text that appears above the form element.
     */
    children?: React.ReactNode;
    /**
     * `id` of the form field that's associated with the `label`. Clicking on the label text will
     * switch the browser's focus to the form field.
     *
     * The Label's `for` and form field's `id` prop must match and be unique to the page. You can
     * use [Lodash's `uniqueId`](https://lodash.com/docs/4.17.10#uniqueId) to generate a unique id.
     */
    for?: string;
    /**
     * Visually disable the label.
     */
    isDisabled?: boolean;
    /**
     * Changes the label text color to red.
     */
    hasError?: boolean;
    /**
     * A selector hook into the React component for use in automated testing environments.
     */
    dataTest?: string;
}

export default function Label({
    isDisabled = false,
    hasError = false,
    for: forProp,
    dataTest,
    children = null,
}: PropTypes): JSX.Element {
    const uiState = getUIState({ isDisabled, hasError });

    return (
        // eslint-disable-next-line jsx-a11y/label-has-for
        <label
            htmlFor={forProp}
            className={classNames({
                [styles.text]: true,
                [styles.textUiStateDisabled]: uiState === 'disabled',
                [styles.textUiStateError]: uiState === 'error',
                [styles.textUiStateDefault]: uiState === 'default',
            })}
            data-test={dataTest}
        >
            {children}
        </label>
    );
}
