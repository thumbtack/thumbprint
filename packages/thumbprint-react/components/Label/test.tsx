import React from 'react';
import { mount } from 'enzyme';
import Label from './index';

test('renders label text', () => {
    const wrapper = mount(<Label for="Duck">Goose</Label>);
    expect(wrapper.text()).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('renders label `for` attribute', () => {
    const wrapper = mount(<Label for="Goose">Foo</Label>);
    expect(wrapper.find('label').prop('htmlFor')).toBe('Goose');
    expect(wrapper).toMatchSnapshot();
});

test('renders disabled label', () => {
    const wrapper = mount(
        <Label isDisabled for="foo">
            Bar
        </Label>,
    );
    expect(wrapper).toMatchSnapshot();
});

test('renders label with error', () => {
    const wrapper = mount(
        <Label hasError for="foo">
            Bar
        </Label>,
    );
    expect(wrapper).toMatchSnapshot();
});

test('adds `dataTest` prop', () => {
    const wrapper = mount(<Label dataTest="Duck">Goose</Label>);
    expect(wrapper.find('label').prop('data-test')).toEqual('Duck');
    expect(wrapper).toMatchSnapshot();
});
