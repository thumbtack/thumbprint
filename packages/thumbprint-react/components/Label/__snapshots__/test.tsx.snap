// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds \`dataTest\` prop 1`] = `
<Label
  dataTest="Duck"
>
  <label
    className="text textUiStateDefault"
    data-test="Duck"
  >
    Goose
  </label>
</Label>
`;

exports[`renders disabled label 1`] = `
<Label
  for="foo"
  isDisabled={true}
>
  <label
    className="text textUiStateDisabled"
    htmlFor="foo"
  >
    Bar
  </label>
</Label>
`;

exports[`renders label \`for\` attribute 1`] = `
<Label
  for="Goose"
>
  <label
    className="text textUiStateDefault"
    htmlFor="Goose"
  >
    Foo
  </label>
</Label>
`;

exports[`renders label text 1`] = `
<Label
  for="Duck"
>
  <label
    className="text textUiStateDefault"
    htmlFor="Duck"
  >
    Goose
  </label>
</Label>
`;

exports[`renders label with error 1`] = `
<Label
  for="foo"
  hasError={true}
>
  <label
    className="text textUiStateError"
    htmlFor="foo"
  >
    Bar
  </label>
</Label>
`;
